File: Baekjoon/B1325.js
const readline = require('readline');

const rl = readline.createInterface({
	input: process.stdin,
	output: process.stdout
});

let input = [];

rl.on('line', function (line){
	input.push(line);
})
.on('close',function(){
	//input에 입력값 한줄씩있음
	let [N, M] = input[0].split(' ').map(el => Number(el));
	let edgeInfo = new Map();
	for (let i=1; i<=M; i++){
		let [child, parent] = input[i].split(' ').map(el => Number(el));
		if (edgeInfo.has(parent))
			edgeInfo.get(parent).push(child);
		else
			edgeInfo.set(parent, [child]);
	}
	console.log(solution(edgeInfo, N));
	process.exit();
});

function solution(edgeInfo, N){
	let max = 0;
	let result = [];
	let isChecked = new Array(N + 1).fill(0);
	for (let i=1; i<=N; i++){
		let cnt = bfs(edgeInfo, isChecked, i);
		if (cnt > max){
			max = cnt;
			result = [i];
		}
		else if (cnt === max)
			result.push(i);
	}
	result.sort((a,b) => a - b);
	return result.join(' ');
}

function bfs(edgeInfo, isChecked, start){
	let queue = [start];
	isChecked[start] = 1;
	let index = 0;
	while (index < queue.length){
		let curr = queue[index];
		if (edgeInfo.has(curr)){
			for (const next of edgeInfo.get(curr)){
				isChecked[next] = 1;
				queue.push(next);
			}
		}
		index++;
	}
	while (queue.length > 0)
		isChecked[queue.pop()] = 0;
	return index;
}

File: Baekjoon/B1325.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class B1325 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int pointCnt = Integer.parseInt(st.nextToken());
        int edgeCnt = Integer.parseInt(st.nextToken());
        ArrayList[] edgeInfo = new ArrayList[pointCnt];
        for (int i=0; i<pointCnt; i++)
            edgeInfo[i] = new ArrayList<Integer>();
        for (int i=0; i<edgeCnt; i++){
            st = new StringTokenizer(br.readLine());
            int child = Integer.parseInt(st.nextToken());
            int parent = Integer.parseInt(st.nextToken());
            edgeInfo[parent - 1].add(child - 1);
        }
        System.out.print(solution(edgeInfo, pointCnt));
    }

    public static String solution(ArrayList[] edgeInfo, int N){
        int max = 0;
        ArrayList<Integer> result = new ArrayList<>();
        for (int i=0; i<N; i++){
            int[] isChecked = new int[N];
            int cnt = dfs(edgeInfo, isChecked, i);
            if (cnt > max){
                result.clear();
                max = cnt;
                result.add(i + 1);
            }
            else if (cnt == max)
                result.add(i + 1);
        }
        StringBuilder sb = new StringBuilder();
        for (int i=0; i<result.size(); i++){
            sb.append(result.get(i));
            if (i != result.size() - 1)
                sb.append(" ");
        }
        return sb.toString();
    }

    public static int dfs(ArrayList[] edgeInfo, int[] isChecked, int curr){
        int cnt = 1;
        isChecked[curr] = 1;
        for (int i=0; i<edgeInfo[curr].size(); i++){
            int next = (int) edgeInfo[curr].get(i);
            if (isChecked[next] == 0)
                cnt += dfs(edgeInfo, isChecked, next);
        }
        return cnt;
    }
}
